CoreUtilities/src/au/gov/asd/tac/constellation/utilities/datastructure/FourTuple.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2010-2021 Australian Signals Directorate
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package au.gov.asd.tac.constellation.utilities.datastructure;
+
+import java.io.Serializable;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+/**
+ * A data-structure for storing a triplet of related values.
+ *
+ * @author arcturus
+ * @author cygnus_x-1
+ *
+ * @param <A> the type of the first object in the 4-tuple
+ * @param <B> the type of the second object in the 4-tuple
+ * @param <C> the type of the third object in the 4-tuple
+ * @param <D> the type of the fourth object in the 4-tuple
+ */
+public class FourTuple<A, B, C, D> implements Serializable, Comparable<FourTuple<A, B, C, D>> {
+
+    private static final long serialVersionUID = 1L;
+
+    private A first;
+    private B second;
+    private C third;
+    private D fourth;
+
+    public FourTuple(final A first, final B second, final C third, final D fourth) {
+        this.first = first;
+        this.second = second;
+        this.third = third;
+        this.fourth = fourth;
+    }
+
+    /**
+     * Convenience method for creating an appropriately typed tuple.
+     *
+     * @param <A> the type of the first object in the 4-tuple
+     * @param <B> the type of the second object in the 4-tuple
+     * @param <C> the type of the third object in the 4-tuple
+     * @param <D> the type of the third object in the 4-tuple
+     * @param first the first object in the tuple
+     * @param second the second object in the tuple
+     * @param third the third object in the tuple
+     * @param fourth the fourth object in the tuple
+     * @return a 4-tuple that is templated with the types of first, second,
+     * third and fourth
+     */
+    public static <A, B, C, D> FourTuple<A, B, C, D> create(A first, B second, C third, D fourth) {
```

Comment:
all the parameters can be final

Reply:
Fixed and updated the other Tuple classes also